### PR TITLE
Feature/move from state bc closes #129

### LINKED ISSUE #129: Function to accept current state + next move in BC
# Task
## Detailed Description

- ?
## Associated story

Story: #127 Claim checkmate, stalemate and surrender and submit to BC/claim ether
## Notes

`moveFromState(gameId, state, signed(state), fromIndex, toIndex, signed(str(fromIndex) + str(toIndex)), player)`
- Verify signature of state
- Verify signature of move
- Apply state, verify move

### PR DESCRIPTION
**Pull request by lisasi**
## Description

The contract now contains a method that gets a signed state and a signed move. It checks whether the signatures are valid and all the mentioned addresses are part of the game. It then applies the state and the move to the game with this gameID after doing some checks.
## Related Issue

#129 Function to accept current state + next move in BC
## Motivation and Context

For off-chain-chess the blockchain acts as a conflict solver and at the end it verifies the end of the game. To do both of these things, we need a way to tell the blockchain the current state of the game. It should then apply the move. If there are any problems with the state or the move the contract should throw and thereby reject this state + move.
## How Has This Been Tested?

Tests in test_contract.js
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
## Checklist contributor:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
## Checklist reviewer:
- [ ] The code has been peer-reviewed by @YourNameHere

### CODE REVIEW

File: app/javascript/cryptoFactory.js [hunk 1]
```diff
@@ -29,7 +59,7 @@ angular.module('dappChess').factory('crypto', function () {
    * @returns{boolean} true, iff the signature matches the account and data
      */
   crypto.verify = function (account, signature, data) {
-    let msgHash = '0x' + web3.sha3(data);
+    let msgHash = web3.solSha3(data);
```
Comment (dennis-westphal) [inline app/javascript/cryptoFactory.js]: why does reference to web3.solSha3 and not just to solSha3?

Comment (lisasi) [inline app/javascript/cryptoFactory.js]: This is fixed in the next commit "Fix hashing in verify" ;-)

**Comment by kyroy** [comment]
Why you want the parameter `opponent`? You can get it by gameId and msg.sender

File: contract/Chess.sol [hunk 2]
```diff
@@ -68,6 +69,70 @@ contract Chess is TurnBasedGame {
...
+
+        if (msg.sender == opponent) {
+            throw;
+        }
+
+        // check whether sender is currentPlayer in state
+        int8 playerColor = msg.sender == gameStates[gameId].playerWhite ? int8(1) : int8(-1);
+
+        //if (state[ChessLogic.Flags(ChessLogic.Flag.CURRENT_PLAYER)] !=  playerColor) {
+        if (state[56] !=  playerColor) {
```
Comment (kyroy) [inline contract/Chess.sol]: Both players should be able to submit the state.

File: contract/Chess.sol [hunk 3]
```diff
@@ -68,6 +69,70 @@ contract Chess is TurnBasedGame {
...
+            throw;
+        }
+
+        /*
+        * Verify signatures
+        */
+        // verify fromIndex
+        if (!verifySig(msg.sender, sha3(fromIndex), sigFromIndex)) {
```
Comment (kyroy) [inline contract/Chess.sol]: Why the user has to sign his own move? He is sending it

File: contract/Chess.sol [hunk 4]
```diff
@@ -68,6 +69,81 @@ contract Chess is TurnBasedGame {
...
+        // find opponent to msg.sender
+        address opponent;
+        if (games[gameId].player1 != msg.sender) {
+            opponent = games[gameId].player1
+        } else {
+            opponent = games[gameId].player1
+        }
+
+        /*
+        * Verify signatures and that mover is current player
```
Comment (kyroy) [inline contract/Chess.sol]: Why the user has to sign his own move? He is sending it

**Comment by kyroy** [comment]
Tests fail:

```
moveFromState()
      1) should accept a valid move with valid signatures

1) Chess contract moveFromState() should accept a valid move with valid signatures:
     AssertionError: expected [Function] to not throw 'Error' but 'Error: Error: VM Exception while executing transaction: invalid JUMP
    at /usr/local/lib/node_modules/ethereumjs-testrpc/lib/blockchain.js:518:18
    [...]
```

**Comment by kyroy** [comment]
Suggestion: https://gist.github.com/Kyroy/6915779ea289db697502d3ee728ac97d

**Comment by graup** [comment]
`games[gameId].nextPlayer = msg.sender` shouldn't this be the opponent? 
Also you have player 1 two times at the top. 
Otherwise I agree, seems good. 
On 6 Jul 2016 10:42 a.m., Dennis Kuhnert notifications@github.com wrote:Suggestion: https://gist.github.com/Kyroy/6915779ea289db697502d3ee728ac97d

**Comment by kyroy** [comment]
Ye, there are some bugs in it. I am fixing them and will commit the changes. The bugs are from this branch...

**Comment by kyroy** [comment]
`games[gameId].nextPlayer = msg.sender` this is sender, because then the `move` function is called ;)

**Pull request merged**